bls12_381/src/pairings.rs

```diff
@@ -558,6 +730,69 @@ fn addition_step(r: &mut G2Projective, q: &G2Affine) -> (Fp2, Fp2, Fp2) {
     (t10, t1, t9)
 }
 
+impl PairingCurveAffine for G1Affine {
+    type Pair = G2Affine;
+    type PairingResult = Gt;
+
+    fn pairing_with(&self, other: &Self::Pair) -> Self::PairingResult {
+        pairing(self, other)
+    }
+}
+
+impl PairingCurveAffine for G2Affine {
+    type Pair = G1Affine;
+    type PairingResult = Gt;
+
+    fn pairing_with(&self, other: &Self::Pair) -> Self::PairingResult {
+        pairing(other, self)
+    }
+}
+
+/// A [`pairing::Engine`] for BLS12-381 pairing operations.
+#[derive(Clone, Debug)]
+pub struct Bls12;
+
+impl Engine for Bls12 {
+    type Fr = Scalar;
+    type G1 = G1Projective;
+    type G1Affine = G1Affine;
+    type G2 = G2Projective;
+    type G2Affine = G2Affine;
```

Comment:
These definitions are incorrect, because pairings rely on G1 and G2 being prime-order groups, but I incorrectly refactored `pairing::Engine` in #241 to use `CofactorCurve` as the bound on these associated types. I'll need to update this PR to provide `G1SubgroupAffine` and `G2SubgroupAffine`, but I'll also figure out whether it's easier to fix `pairing::Engine` before this PR, or wait until after we've dropped the old implementations.

Reply:
I refactored `pairing::Engine` in #268 and rebased this PR to remove `G1Subgroup` and `G2Subgroup`.